Fix: don't validate schemas for disabled rules (fixes #7690)

not-an-aardvark opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix

See https://github.com/eslint/eslint/issues/7690

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

This updates `config-validator.js` to avoid option validation when a rule is disabled. This ensures that `some-rule: off` is always a valid way to disable a rule, and it doesn't throw a schema error if the rule requires options when enabled.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

gyandeeps commented:
LGTM. 

**Feedback:** I think from a functional perspective, every function has a responsibility. In this case its responsibility is to validate a rule's config. I thinking mixing the logic of if the rule is disabled, dont valuate the config is some how not right. It takes away the reusability or purpose of the function. I am of the mind set to only call this function when you really want to check config.

Again this should not be a PR blocker but something I thought about.

not-an-aardvark commented:
@gyandeeps From what I can tell, the function currently validates both rule severity and rule options. I see your point about the single-responsibility principle, but I'm not sure I agree that this function violates it, because its responsibility is basically "ensure that the stuff passed to the rule is valid". Depending on the value of the severity, this could either mean "check that the severity is valid" or "check that the severity and the rule schema are both valid". In other words, I think it's reasonable to put the check in this function because it's directly related to the function's behavior.

ilyavolodin commented:
I think I agree with @gyandeeps. I think a better way would be not to call `validate` or `validateRuleOptions` functions if the rule is off.

platinumazure commented:
I think I agree with @gyandeeps and @ilyavolodin. Maybe it would also be better to separate validation of rule severity and rule options into their own functions as well, to ensure single responsibility and allow more control of how we handle invalid results in those areas.

not-an-aardvark commented:
Updated to follow @platinumazure's suggestion; I separated severity validation and rule-specific options validation into separate functions.

Since the `validateRuleOptions` function is part of `config-validator`'s API, I decided to keep that function the same, but I split its functionality into `validateRuleSeverity` and `validateRuleSchema`.

The only user-visible effect of this change is that if a rule's severity *and* options are invalid, ESLint will only report the invalid severity, rather than reporting both the invalid severity and the invalid options. This is because the validator can't determine whether it should validate the options if the schema is invalid.

platinumazure commented on tests/lib/config/config-validator.js:
```diff
@@ -103,6 +120,12 @@ describe("Validator", () => {
+            const fn = validator.validate.bind(null, { rules: { "mock-required-options-rule": "off" } }, "tests");
+
+            assert.doesNotThrow(fn);
+        });
```
Just to be absolutely sure, can we have a test with `["off"]` as the configured set of options? If nothing else, it could prevent a regression later.

btmills commented:
This gets a pretty nice perf bump too:

```sh
$ npm run perf

> eslint@3.11.1 perf /Users/brandon/code/eslint/eslint
> node Makefile.js perf

Downloading the repository of multi-files performance test target.

Loading:
  Load performance Run #1:  259.611759ms
  Load performance Run #2:  208.566536ms
  Load performance Run #3:  209.223229ms
  Load performance Run #4:  206.482016ms
  Load performance Run #5:  205.450401ms

  Load Performance median:  208.566536ms


Single File:
  CPU Speed is 2200 with multiplier 13000000
  Performance Run #1:  6095.261969ms
  Performance Run #2:  6284.323343ms
  Performance Run #3:  6202.953062ms
  Performance Run #4:  6180.078094ms
  Performance Run #5:  6324.969696ms

  Performance budget exceeded: 6202.953062ms (limit: 5909.090909090909ms)


Multi Files (0 files):
  CPU Speed is 2200 with multiplier 39000000
  Performance Run #1:  15290.355138ms
  Performance Run #2:  15229.238735ms
  Performance Run #3:  15637.211067ms
  Performance Run #4:  15118.806527ms
  Performance Run #5:  15465.255259ms

  Performance budget ok:  15290.355138ms (limit: 17727.272727272728ms)

$ git co disabled-config-validation
Branch disabled-config-validation set up to track remote branch disabled-config-validation from origin.
Switched to a new branch 'disabled-config-validation'
$ npm run perf

> eslint@3.11.1 perf /Users/brandon/code/eslint/eslint
> node Makefile.js perf


Loading:
  Load performance Run #1:  217.830524ms
  Load performance Run #2:  200.390492ms
  Load performance Run #3:  212.837244ms
  Load performance Run #4:  200.300796ms
  Load performance Run #5:  210.901808ms

  Load Performance median:  210.901808ms


Single File:
  CPU Speed is 2200 with multiplier 13000000
  Performance Run #1:  5955.571386ms
  Performance Run #2:  5967.169093ms
  Performance Run #3:  6106.626481ms
  Performance Run #4:  5919.5527170000005ms
  Performance Run #5:  6004.485014ms

  Performance budget exceeded: 5967.169093ms (limit: 5909.090909090909ms)


Multi Files (0 files):
  CPU Speed is 2200 with multiplier 39000000
  Performance Run #1:  14594.376572ms
  Performance Run #2:  14817.782568ms
  Performance Run #3:  14549.666905ms
  Performance Run #4:  14501.114731ms
  Performance Run #5:  14691.406546ms

  Performance budget ok:  14594.376572ms (limit: 17727.272727272728ms)
```
